woocommerce/woocommerce
# Add/e2e implement describe each

### Changes proposed in this Pull Request:

- This PR implements `describe.each` and `it.each` on tests that have repetitive steps but use different test data on each iteration. This will not improve performance that much, but will make our tests more concise and readable.
- It contains necessary cherry-picked commits from #30184.
- I refactored [this variable](https://github.com/woocommerce/woocommerce/pull/30119/files#diff-4c0e5f5a585ade4c9bf11d9c57b30abf211cbf3be265168d6017ebd7b7945cd3L15-R23) to make it easier to access its values from the `each`'ed test.

Closes #29678.

### How to test the changes in this Pull Request:

1. Set up e2e environment by running the onboarding & activation tests:
    ```
    npx wc-e2e test:e2e ./tests/e2e/specs/activate-and-setup/setup-onboarding.js
    ```
2. Run the "merchant order search" test:
    ```
    npx wc-e2e test:e2e-dev ./tests/e2e/specs/wp-admin/order-searching.test.js
    ```
3. All these tests should pass:
    <img width="657" alt="Screen Shot 2021-06-17 at 3 45 14 PM" src="https://user-images.githubusercontent.com/4509348/122353891-0f57a680-cf83-11eb-8961-5bf05be40350.png">
4. Run the other tests where we implemented `each()`:
    1. `E2E_RETEST=1 npx wc-e2e test:e2e ./tests/e2e/specs/activate-and-setup/onboarding-tasklist.test.js`
        <img width="459" alt="Screen Shot 2021-07-01 at 2 24 20 PM" src="https://user-images.githubusercontent.com/4509348/124263438-d401ce00-db65-11eb-8856-a2ad6d304285.png">
    1. `npx wc-e2e test:e2e ./tests/e2e/specs/wp-admin/analytics-page-loads.test.js`
        <img width="467" alt="Screen Shot 2021-07-01 at 3 29 28 PM" src="https://user-images.githubusercontent.com/4509348/124263804-4d99bc00-db66-11eb-8ae3-e452188ab8d8.png">
    1. `npx wc-e2e test:e2e ./tests/e2e/specs/wp-admin/order-status-filters.test.js`
        <img width="464" alt="Screen Shot 2021-07-02 at 6 53 25 PM" src="https://user-images.githubusercontent.com/4509348/124264273-d87ab680-db66-11eb-92b4-70ca8fe39a1c.png">
    1. `npx wc-e2e test:e2e ./tests/e2e/specs/front-end/cart-coupons.test.js`
        <img width="465" alt="Screen Shot 2021-07-01 at 7 22 34 PM" src="https://user-images.githubusercontent.com/4509348/124264460-1546ad80-db67-11eb-8a8c-3a7bc7ad4744.png">
    1. `npx wc-e2e test:e2e ./tests/e2e/specs/front-end/checkout-coupons.test.js`
        <img width="466" alt="Screen Shot 2021-07-01 at 7 32 31 PM" src="https://user-images.githubusercontent.com/4509348/124264586-3e673e00-db67-11eb-9826-5c4df0641f3e.png">
    1. `npx wc-e2e test:e2e ./tests/e2e/specs/front-end/my-account.test.js`
        <img width="506" alt="Screen Shot 2021-07-01 at 7 49 20 PM" src="https://user-images.githubusercontent.com/4509348/124264696-66ef3800-db67-11eb-9039-40b4dfee4ebc.png">




        



### Other information:

### Changelog entry

> Implemented data-driven design using `describe.each` and `it.each`
